### PR TITLE
Telegram Notifications Support

### PR DESCRIPTION
**Pull request by marbink**
Add Notifications on Telegram.
Add two config entry to set bot token and recipient chat_id.

### CODE REVIEW

File: notification.py [hunk 1]
```diff
@@ -281,8 +282,43 @@ def notify(self):
...
+                'title' : title,
+                'address' : description,
+            }
+
+            r = requests.get(TELEGRAM_BASE_URL.format(token=config.TELEGRAM_BOT_TOKEN, command = cmd), params=payload)
+            if r.status_code == 200:
+                self.logger.info('Sent a Telegram notification about {}.'.format(self.name))
+                return True
+            else:
+                self.logger.info('Something goes wrong with sending a notification to Telegram about {}.'.format(self.name))
```
Comment (Noctem) [inline notification.py]: How about:
`'Failed to send a Telegram notification about {}.'`

File: notification.py [hunk 2]
```diff
@@ -281,8 +282,43 @@ def notify(self):
...
-        return tweeted or pushed
+        return tweeted or pushed or telegram
+
+    def sendToTelegram(self):
+        try:
+            TELEGRAM_BASE_URL = "https://api.telegram.org/bot{token}/{command}"
+            title = self.name
+            if self.expire_time:
+                minutes, seconds = divmod(self.delta.total_seconds(), 60)
+                description = 'Expire at: {e} ({m}m{s:.0f}s left)'.format(e=self.expire_time, m=int(minutes), s=seconds)
```
Comment (Noctem) [inline notification.py]: Expire**s**

File: notification.py [hunk 3]
```diff
@@ -281,8 +282,43 @@ def notify(self):
...
+            else:
+                description = ("It'll expire between {e1} & {e2}.").format(e1=self.min_expire_time, e2=self.max_expire_time)
+            
+            if self.iv[0] is not None:
+                title += ' ({a}/{d}/{s})'.format(a=self.iv[0], d=self.iv[1], s=self.iv[2])
```
Comment (Noctem) [inline notification.py]:
```python
title += ' ({0[0]}/{0[1]}/{0[2]})'.format(self.iv)
```
Would be more concise and would [keep the line under 79 characters](https://www.python.org/dev/peps/pep-0008/#maximum-line-length), which I usually try to do if it's not too much hassle.

File: notification.py [hunk 4]
```diff
@@ -36,8 +37,8 @@
         setattr(config, setting_name, default)
 
 if config.NOTIFY:
-    if not (config.PB_API_KEY or config.TWITTER_ACCESS_KEY):
-        raise ValueError('NOTIFY is enabled but no PushBullet or Twitter keys were provided.')
+    if not (config.PB_API_KEY or config.TWITTER_ACCESS_KEY or (config.TELEGRAM_BOT_TOKEN and config.TELEGRAM_CHAT_ID)):
+        raise ValueError('NOTIFY is enabled but no PushBullet or Twitter keys or Telegram Token & Chat ID were provided.')
```
Comment (Noctem) [inline notification.py]: How about:
```
'NOTIFY is enabled but no Twitter, Telegram, or PushBullet keys were provided.'
```

File: config.example.py [hunk 5]
```diff
@@ -216,13 +216,20 @@
 ### The following values are fake, replace them with your own keys to enable
 ### PushBullet notifications and/or tweeting, otherwise leave them out of your
 ### config or set them to None.
-## you must provide keys for at least one service (Twitter and/or PushBullet) to use notifications
+## you must provide keys for at least one service (Twitter and/or PushBullet and/or Telegram) to use notifications
```
Comment (Noctem) [inline config.example.py]:
```
(PushBullet, Telegram, and/or Twitter)
```

File: config.example.py [hunk 6]
```diff
@@ -216,13 +216,20 @@
...
+## Telegram bot token is the one Botfather is sending to you after completing bot creation.
```
Comment (Noctem) [inline config.example.py]: `sends` instead of `is sending`

**Comment by marbink** [comment]
@Noctem Changed labels with the ones you requested.
About coordinates, Telegram is already handling them: clicking on the message, a new tab will appear containing a map (Apple Map based on iOS, GMaps on Android). On Desktop GMaps will be launched in a new tab.

About images, bots can send images. But I need to check if it is possibile to merge image+coords+text in one message. Otherwise, sending multiple messages for each monster, we could get some unexpected behaviour, for example: [IMG_Pok1] [IMG_Pok2] [Location1] [Location2]. There is no assurance about the receiving order. I don't think about this for now. I'll take a look in future, btw.

**Comment by Noctem** [comment]
@marbink Cool, I wish Twitter would do that. Their official clients hide the exact tweet coordinates.

Good to know. It's likely best to just skip images on Telegram if it's not possible to include them all in the same notification.

**Comment by thordurk91** [comment]
Telegram in the kvangent PokeAlarm sends 3 seperate messages for each notifications. One with text about pokemon, one a sticker image for the pokemon and one for the google maps link.

The order can be confused when alerting for 2 pokemon at the same time, or by a delay in the webhook. We could maybe solve this by linking an increasing number/ID to the notification and have it with the pokemon image and the later message i.e. google map link

**Comment by Noctem** [comment]
@thordurk91 Hmm, multiple messages per notification seems a bit excessive.

**Comment by thordurk91** [comment]
@Noctem I agree 3 is excessive and clutters the pings so they cant quickview it in their notifications. 

The single link is superior in terms of usability, but a good addition might be adding the zone defined in the config to the string that telegram shows with the google maps link. Since the map is useful but rather small. 

How my telegram looks like:
screenshot_20170107-173356 1

**Comment by Noctem** [comment]
@thordurk91 Is it possible to customize the size or zoom level? I don't see any options for that in [their documentation](https://core.telegram.org/bots/api#sendvenue).

Also, for anyone using (or intending to use) this, would you prefer the current format (with just the expiration time in the message body) or something more like the [Twitter](https://twitter.com/SLCPokemon) and PushBullet notifications (with a description of the location if Landmarks are set).

**Comment by thordurk91** [comment]
@Noctem I looked a bit into it, and was running into my notification not being posted when I used sendMessage instead of sendVenue(current).

There's nothing to change the default zoom on venue from the looks of it, but maybe if you define the zoom in a google maps link in a message it might display it at the designated zoom.

Something like this, where the 8z is the zoom level. But I haven't gotten it to work yet. How do I turn it on so I can see the logs/expections from notifications.py, if they arent printed out by default?

            payload = {
                'chat_id': config.TELEGRAM_CHAT_ID,
                'text': '<b' + title + '</b>' + description + '\n <a href="https://www.google.com/maps/preview/@' + str(self.coordinates[0]) + ',' + str(self.coordinates[1]) + ',8z">Map Link</a>',
                'parse_mode': 'HTML',
                'disable_web_page_preview': 'False',
                'disable_notification': 'False'
            }

**Comment by GouSao** [comment]
@thordurk91 is very good to merge if included IV encounter code.

**Comment by dcarpi** [comment]
@Noctem Hello, I can't figure out how to manage the notification via Telegram, maybe the webhook is not installed ? how to install ?

thanks

**Pull request merged**